File: FBDeviceControl/Video/FBDeviceVideo.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDeviceVideo.h"

#import <FBControlCore/FBControlCore.h>

#import <AVFoundation/AVFoundation.h>

#import <CoreMediaIO/CMIOHardwareObject.h>
#import <CoreMediaIO/CMIOHardwareSystem.h>
#import <CoreFoundation/CoreFoundation.h>

#import "FBDevice.h"
#import "FBDeviceControlError.h"
#import "FBDeviceControlFrameworkLoader.h"
#import "FBDeviceVideoFileEncoder.h"

@interface FBDeviceVideo ()

@property (nonatomic, strong, readonly) FBDeviceVideoFileEncoder *encoder;
@property (nonatomic, strong, readonly) dispatch_queue_t workQueue;

@end

@implementation FBDeviceVideo

#pragma mark Initialization Helpers

+ (FBFuture<AVCaptureDevice *> *)findCaptureDeviceForDevice:(FBDevice *)device
{
  return [[FBFuture
    onQueue:device.workQueue resolveUntil:^{
      AVCaptureDevice *captureDevice = [AVCaptureDevice deviceWithUniqueID:device.udid];
      if (!captureDevice) {
        return [[[FBDeviceControlError
          describeFormat:@"Capture Device %@ not available", device.udid]
          noLogging]
          failFuture];
      }
      return [FBFuture futureWithResult:captureDevice];
    }]
    timeout:FBControlCoreGlobalConfiguration.fastTimeout waitingFor:@"Device %@ to have an associated capture device appear", device];
}

+ (BOOL)allowAccessToScreenCaptureDevicesWithError:(NSError **)error
{
  CMIOObjectPropertyAddress properties = {
    kCMIOHardwarePropertyAllowScreenCaptureDevices,
    kCMIOObjectPropertyScopeGlobal,
    kCMIOObjectPropertyElementMaster,
  };
  UInt32 allow = 1;
  OSStatus status = CMIOObjectSetPropertyData(
    kCMIOObjectSystemObject,
    &properties,
    0,
    NULL,
    sizeof(allow),
    &allow
  );
  if (status != 0) {
    return [[FBDeviceControlError
      describeFormat:@"Failed to enable Screen Capture devices with status %d", status]
      failBool:error];
  }
  return YES;
}

#pragma mark Initializers

+ (FBFuture<AVCaptureSession *> *)captureSessionForDevice:(FBDevice *)device
{
  // Allow Access
  NSError *error = nil;
  if (![self allowAccessToScreenCaptureDevicesWithError:&error]) {
    return [FBFuture futureWithError:error];
  }
  // Obtain the Capture Device
  return [[self
    findCaptureDeviceForDevice:device]
    onQueue:device.workQueue fmap:^(AVCaptureDevice *captureDevice) {
      // Get the Input instance for this Device.
      NSError *innerError = nil;
      AVCaptureDeviceInput *deviceInput = [AVCaptureDeviceInput deviceInputWithDevice:captureDevice error:&innerError];
      if (!deviceInput) {
        return [[[FBDeviceControlError
          describeFormat:@"Failed to create Device Input for %@", captureDevice]
          causedBy:innerError]
          failFuture];
      }
      // Add the Input to a new Session.
      AVCaptureSession *session = [[AVCaptureSession alloc] init];
      if (![session canAddInput:deviceInput]) {
        return [[FBDeviceControlError
          describeFormat:@"Cannot add Device Input to session for %@", captureDevice]
          failFuture];
      }
      [session addInput:deviceInput];

      return [FBFuture futureWithResult:session];
    }];
}

+ (FBFuture<FBDeviceVideo *> *)videoForDevice:(FBDevice *)device filePath:(NSString *)filePath
{
  // Add the Input to a new Session.
  return [[self
    captureSessionForDevice:device]
    onQueue:device.workQueue fmap:^(AVCaptureSession *session) {
      // Construct the Device Video instance.
      NSError *error = nil;
      FBDeviceVideoFileEncoder *encoder = [FBDeviceVideoFileEncoder encoderWithSession:session filePath:filePath logger:device.logger error:&error];
      if (!encoder) {
        return [FBFuture futureWithError:error];
      }
      FBDeviceVideo *video = [[FBDeviceVideo alloc] initWithEncoder:encoder workQueue:device.workQueue];
      return [FBFuture futureWithResult:video];
    }];
}

- (instancetype)initWithEncoder:(FBDeviceVideoFileEncoder *)encoder workQueue:(dispatch_queue_t)workQueue
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _encoder = encoder;
  _workQueue = workQueue;

  return self;
}

#pragma mark Public

- (FBFuture<NSNull *> *)startRecording
{
  return [self.encoder startRecording];
}

- (FBFuture<NSNull *> *)stopRecording
{
  return [self.encoder stopRecording];
}

#pragma mark FBiOSTargetContinuation

- (FBiOSTargetFutureType)futureType
{
  return FBiOSTargetFutureTypeVideoRecording;
}

- (FBFuture<NSNull *> *)completed
{
  FBDeviceVideoFileEncoder *encoder = self.encoder;
  return [[encoder
    completed]
    onQueue:self.workQueue respondToCancellation:^{
      return [encoder stopRecording];
    }];
}

@end

File: FBDeviceControl/Video/FBDeviceVideo.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import AVFoundation
import CoreMediaIO
@preconcurrency import FBControlCore
import Foundation

protocol FBDeviceVideoCaptureDeviceCandidate {
  var uniqueID: String { get }
  var localizedName: String { get }
  var modelID: String { get }

  func hasMediaType(_ mediaType: AVMediaType) -> Bool
}

extension AVCaptureDevice: FBDeviceVideoCaptureDeviceCandidate {}

public final class FBDeviceVideo {
  private let encoder: FBVideoFileWriter
  private let workQueue: DispatchQueue

  private static let captureDeviceDiscoveryPollNanoseconds: UInt64 = 250_000_000

  // MARK: Initialization Helpers

  private class func videoCaptureAuthorizationStatusString(_ status: AVAuthorizationStatus) -> String {
    switch status {
    case .notDetermined:
      return "notDetermined"
    case .restricted:
      return "restricted"
    case .denied:
      return "denied"
    case .authorized:
      return "authorized"
    @unknown default:
      return "unknown(\(status.rawValue))"
    }
  }

  private class func videoCaptureAuthorizationError(_ status: AVAuthorizationStatus) -> Error {
    if status == .denied {
      return FBDeviceControlError.describe(
        "Camera authorization is denied. Grant camera access to the application responsible for idb_companion in System Settings > Privacy & Security > Camera"
      ).build()
    }
    if status == .restricted {
      return FBDeviceControlError.describe("Camera authorization is restricted by system policy").build()
    }
    return FBDeviceControlError.describe(
      "Camera authorization request did not grant access (status: \(videoCaptureAuthorizationStatusString(status)))"
    ).build()
  }

  static func ensureVideoCaptureAuthorization(
    logger: any FBControlCoreLogger,
    authorizationStatus: () -> AVAuthorizationStatus,
    requestAccess: () async -> Bool
  ) async throws {
    let status = authorizationStatus()
    logger.log("[FBDeviceVideo] Camera authorization status: \(videoCaptureAuthorizationStatusString(status))")

    switch status {
    case .authorized:
      return
    case .denied, .restricted:
      throw videoCaptureAuthorizationError(status)
    case .notDetermined:
      let granted = await requestAccess()
      let finalStatus = authorizationStatus()
      logger.log(
        "[FBDeviceVideo] Camera authorization status after request: \(videoCaptureAuthorizationStatusString(finalStatus)) (granted=\(granted ? "YES" : "NO"))"
      )
      guard granted, finalStatus == .authorized else {
        throw videoCaptureAuthorizationError(finalStatus)
      }
    @unknown default:
      throw videoCaptureAuthorizationError(status)
    }
  }

  private class func allowAccessToScreenCaptureDevices() throws {
    var properties = CMIOObjectPropertyAddress(
      mSelector: CMIOObjectPropertySelector(kCMIOHardwarePropertyAllowScreenCaptureDevices),
      mScope: CMIOObjectPropertyScope(kCMIOObjectPropertyScopeGlobal),
      mElement: CMIOObjectPropertyElement(kCMIOObjectPropertyElementMain)
    )
    var allow: UInt32 = 1
    let status = CMIOObjectSetPropertyData(
      CMIOObjectID(kCMIOObjectSystemObject),
      &properties,
      0,
      nil,
      UInt32(MemoryLayout<UInt32>.size),
      &allow
    )
    if status != 0 {
      throw FBDeviceControlError.describe("Failed to enable Screen Capture devices with status \(status)").build()
    }
  }

  private class func normalizedIdentifier(_ identifier: String) -> String {
    identifier
      .components(separatedBy: CharacterSet.alphanumerics.inverted)
      .joined()
      .lowercased()
  }

  static func captureDevice<Candidate: FBDeviceVideoCaptureDeviceCandidate>(
    forUDID targetUDID: String,
    named targetName: String,
    connectedDeviceUDIDs: [String],
    from candidates: [Candidate]
  ) -> Candidate? {
    let normalizedUDID = normalizedIdentifier(targetUDID)
    var screenCandidates: [Candidate] = []
    var identifierMatches: [Candidate] = []
    var nameMatches: [Candidate] = []

    for candidate in candidates {
      guard candidate.modelID == "iOS Device", candidate.hasMediaType(.muxed) else {
        continue
      }
      screenCandidates.append(candidate)

      let normalizedCandidateID = normalizedIdentifier(candidate.uniqueID)
      if !normalizedUDID.isEmpty, normalizedCandidateID.contains(normalizedUDID) {
        identifierMatches.append(candidate)
      }
      if candidate.localizedName == targetName {
        nameMatches.append(candidate)
      }
    }

    if !identifierMatches.isEmpty {
      return identifierMatches.count == 1 ? identifierMatches[0] : nil
    }
    if !nameMatches.isEmpty {
      return nameMatches.count == 1 ? nameMatches[0] : nil
    }
    if screenCandidates.count == 1,
       connectedDeviceUDIDs.count == 1,
       connectedDeviceUDIDs[0] == targetUDID
    {
      return screenCandidates[0]
    }
    return nil
  }

  private class func captureDeviceInventory(_ candidates: [AVCaptureDevice]) -> String {
    let descriptions = candidates.map { candidate in
      var mediaTypes: [String] = []
      if candidate.hasMediaType(.video) {
        mediaTypes.append("video")
      }
      if candidate.hasMediaType(.audio) {
        mediaTypes.append("audio")
      }
      if candidate.hasMediaType(.muxed) {
        mediaTypes.append("muxed")
      }
      return "{name=\(candidate.localizedName), uniqueID=\(candidate.uniqueID), modelID=\(candidate.modelID), deviceType=\(candidate.deviceType.rawValue), mediaTypes=\(mediaTypes.joined(separator: ","))}"
    }.sorted()
    return descriptions.isEmpty ? "<none>" : descriptions.joined(separator: "; ")
  }

  private class func captureDeviceDiscoverySession() -> AVCaptureDevice.DiscoverySession {
    if #available(macOS 14.0, *) {
      return AVCaptureDevice.DiscoverySession(
        deviceTypes: [.external, .continuityCamera],
        mediaType: nil,
        position: .unspecified
      )
    }
    return AVCaptureDevice.DiscoverySession(
      deviceTypes: [.externalUnknown],
      mediaType: nil,
      position: .unspecified
    )
  }

  private class func findCaptureDeviceAsync(for device: FBDevice) async throws -> AVCaptureDevice {
    let timeout = FBControlCoreGlobalConfiguration.fastTimeout
    let deadline = Date().addingTimeInterval(timeout)
    let logger = device.logger ?? FBControlCoreGlobalConfiguration.defaultLogger
    var discoverySession: AVCaptureDevice.DiscoverySession?
    var lastInventory: String?

    while true {
      if let captureDevice = AVCaptureDevice(uniqueID: device.udid) {
        return captureDevice
      }

      if discoverySession == nil {
        discoverySession = captureDeviceDiscoverySession()
      }
      let candidates = discoverySession?.devices ?? []
      let inventory = captureDeviceInventory(candidates)
      if inventory != lastInventory {
        logger.log("[FBDeviceVideo] Capture device candidates: \(inventory)")
        lastInventory = inventory
      }

      let connectedDeviceUDIDs = device.set?.allDevices.map(\.udid) ?? []
      if let captureDevice = captureDevice(
        forUDID: device.udid,
        named: device.name,
        connectedDeviceUDIDs: connectedDeviceUDIDs,
        from: candidates
      ) {
        logger.log(
          "[FBDeviceVideo] Matched screen capture device name=\(captureDevice.localizedName) uniqueID=\(captureDevice.uniqueID) modelID=\(captureDevice.modelID)"
        )
        return captureDevice
      }

      if Date() >= deadline {
        throw FBDeviceControlError.describe("Timed out waiting \(timeout)s for device \(device) to have an associated capture device appear").build()
      }
      try await Task.sleep(nanoseconds: captureDeviceDiscoveryPollNanoseconds)
    }
  }

  // MARK: Initializers

  public class func captureSessionAsync(for device: FBDevice) async throws -> AVCaptureSession {
    try await captureSessionAsync(
      logger: device.logger ?? FBControlCoreGlobalConfiguration.defaultLogger,
      authorizationStatus: {
        AVCaptureDevice.authorizationStatus(for: .video)
      },
      requestAccess: {
        await AVCaptureDevice.requestAccess(for: .video)
      },
      allowAccessToScreenCaptureDevices: {
        try allowAccessToScreenCaptureDevices()
      },
      findCaptureDevice: {
        try await findCaptureDeviceAsync(for: device)
      }
    )
  }

  static func captureSessionAsync(
    logger: any FBControlCoreLogger,
    authorizationStatus: () -> AVAuthorizationStatus,
    requestAccess: () async -> Bool,
    allowAccessToScreenCaptureDevices: () throws -> Void,
    findCaptureDevice: () async throws -> AVCaptureDevice
  ) async throws -> AVCaptureSession {
    try await ensureVideoCaptureAuthorization(
      logger: logger,
      authorizationStatus: authorizationStatus,
      requestAccess: requestAccess
    )
    try allowAccessToScreenCaptureDevices()
    let captureDevice = try await findCaptureDevice()
    let deviceInput = try AVCaptureDeviceInput(device: captureDevice)
    let session = AVCaptureSession()
    if !session.canAddInput(deviceInput) {
      throw FBDeviceControlError.describe("Cannot add Device Input to session for \(captureDevice)").build()
    }
    session.addInput(deviceInput)
    return session
  }

  public class func videoAsync(for device: FBDevice, filePath: String) async throws -> FBDeviceVideo {
    let session = try await captureSessionAsync(for: device)
    let encoder = try FBVideoFileWriter.writer(withSession: session, filePath: filePath, logger: device.logger ?? FBControlCoreGlobalConfiguration.defaultLogger)
    return FBDeviceVideo(encoder: encoder, workQueue: device.workQueue)
  }

  private init(encoder: FBVideoFileWriter, workQueue: DispatchQueue) {
    self.encoder = encoder
    self.workQueue = workQueue
  }

  // MARK: Public

  public func startRecording() -> FBFuture<NSNull> {
    encoder.startRecording()
  }

  public func stopRecording() -> FBFuture<NSNull> {
    encoder.stopRecording()
  }
}
